urlc8dbc7ce6e_dlubarov_guava_tgz-pJ8-MainJ8/src/d/TypeDef.java

```diff
@@ -0,0 +1,139 @@
+            virtualFieldTable.put(allInstanceFields[i].intern(), i);
+
+        this.virtualMethodDescTable = virtualMethodDescTable;
+        virtualMethodTable = new HashMap<MethodDef, ConcreteMethodDef>();
+
+        this.superGenericDescs = superGenericDescs;
+        superGenerics = new HashMap<TypeDef, SuperType[]>();
+
+        if (genericVariances == null || genericVariances.length == 0)
+            typeNoGenerics = new ConcreteType(this);
+
+        God.newType(this);
+    }
+
+    public void link() {
+        for (MethodDef method : staticMethods)
+            method.link();
+        for (MethodDef method : instanceMethods)
+            method.link();
+
+        // Unless this is an abstract type, populate the virtual method table.
+        if (virtualMethodDescTable != null)
+            for (RawMethod aDesc : virtualMethodDescTable.keySet()) {
+                RawMethod bDesc = virtualMethodDescTable.get(aDesc);
+                MethodDef a = God.resolveMethod(aDesc), b = God.resolveMethod(bDesc);
+                if (!(b instanceof ConcreteMethodDef))
+                    throw new RuntimeException("vtable points to abstract method " + b);
+                virtualMethodTable.put(a, (ConcreteMethodDef) b);
+            }
+
+        // Populate the supertype generics table.
+        for (RawType superDesc : superGenericDescs.keySet()) {
+            TypeDesc[] superArgDescs = superGenericDescs.get(superDesc);
+            TypeDef supertype = God.resolveType(superDesc);
+            SuperType[] superArgs = new SuperType[superArgDescs.length];
+            for (int i = 0; i < superArgs.length; ++i)
+                superArgs[i] = superArgDescs[i].toSuper();
+            superGenerics.put(supertype, superArgs);
+        }
+    }
+
+    public void init() {
+        if (initializationStatus == 2)
+            return;
+        if (initializationStatus == 1)
+            throw new NiftyException("Circular dependency exists in static initializers.");
+
+        initializationStatus = 1;
+        for (ConcreteMethodDef m : staticMethods)
+            if (m.desc.equals(new RawMethod(true, desc, "init", 0, TypeDesc.NONE)))
+                m.invoke(ConcreteType.NONE);
+        initializationStatus = 2;
+    }
+
+    public BaseObject rawInstance(ConcreteType type) {
+        return new NormalObject(type);
+    }
+
+    public BaseObject rawInstanceNoGenerics() {
+        assert this.genericVariances.length == 0;
+        return new NormalObject(typeNoGenerics);
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
+        sb.append("type ").append(desc).append(" {");
+
+        // Unless abstract, append the virtual method descriptor table.
+        if (virtualMethodDescTable != null) {
+            sb.append("\n    Virtual method desc table:\n");
+            for (RawMethod a : virtualMethodDescTable.keySet()) {
+                RawMethod b = virtualMethodDescTable.get(a);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.